src/main/java/net/fabricmc/fabric/impl/listener/ListenerTypeFactoryImpl.java

```diff
@@ -0,0 +1,101 @@
+package net.fabricmc.fabric.impl.listener;
+
+import net.fabricmc.fabric.api.event.listener.ListenerChainer;
+import net.fabricmc.fabric.api.event.listener.ListenerType;
+import net.fabricmc.fabric.api.event.listener.ListenerTypeFactory;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Array;
+import java.util.*;
+
+public class ListenerTypeFactoryImpl implements ListenerTypeFactory {
+	static class Type<T> extends ListenerType<T> {
+		private final Class<T> type;
+		private final List<Object> listeners = new ArrayList<>();
+
+		Type(Class<T> type) {
+			this.type = type;
+			update();
+		}
+
+		@SuppressWarnings("unchecked")
+		private void update() {
+			switch (listeners.size()) {
+				case 0:
+					reference = (T) ListenerTypeFactoryImpl.INSTANCE.emptyRefMap.get(type);
+					break;
+				case 1:
+					reference = (T) listeners.get(0);
+					break;
+				default:
+					ListenerChainer<?> chainer = ListenerTypeFactoryImpl.INSTANCE.listenerChainerMap.get(type);
+					if (chainer == null) {
+						reference = null;
+					} else {
+
+						Object array = Array.newInstance(type, listeners.size());
+						for (int i = 0; i < listeners.size(); i++) {
+							Array.set(array, i, listeners.get(i));
+						}
+
```

Comment:
Making sure I understand this: are you copying here as a defensive measure? Is it simply to prevent modification?  Doesn't seem like you're trying to support concurrency overall.

Reply:
Java arrays are constant-size, so I have to create a new one every time a listener is registered. But their iteration is faster.

Reply:
Ah, okay. This “update” is the thing that builds the handler, not the handler itself. This makes much more sense now and matches what I was thinking - just wasn’t seeing it. I guess my only critique at this point is the naming didn’t help me understand it, but I can’t offer any suggestions atm.

Reply:
Ah, okay. This “update” is the thing that builds the handler, not the handler itself. This makes much more sense now and matches what I was thinking - just wasn’t seeing it. I guess my only critique at this point is the naming didn’t help me understand it, but I can’t offer any suggestions atm.